luxtronik/__init__.py

```diff
@@ -74,43 +74,60 @@ def __del__(self):
                 "Disconnected from Luxtronik heatpump %s:%s", self._host, self._port
             )
 
-    def read(self):
-        """Read data from heatpump."""
-        return self._read_after_write(parameters=None)
-
-    def write(self, parameters):
-        """Write parameter to heatpump."""
-        return self._read_after_write(parameters=parameters)
-
-    def _read_after_write(self, parameters):
+    def _connect(self, func):
         """
-        Read and/or write value from and/or to heatpump.
+        Decorator around various read/write functions to connect first.
+
         This method is essentially a wrapper for the _read() and _write()
         methods.
         Locking is being used to ensure that only a single socket operation is
         performed at any point in time. This helps to avoid issues with the
         Luxtronik controller, which seems unstable otherwise.
-        If write is true, all parameters will be written to the heat pump
-        prior to reading back in all data from the heat pump. If write is
-        false, no data will be written, but all available data will be read
-        from the heat pump.
+        """
+
+        def inner(*args, **kwargs):
+            with self._lock:
+                is_none = self._socket is None
+                if is_none:
+                    self._socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+                if is_none or is_socket_closed(self._socket):
+                    self._socket.connect((self._host, self._port))
+                    LOGGER.info(
+                        "Connected to Luxtronik heatpump %s:%s", self._host, self._port
```

Comment:
We should agree on a spelling here, i.e. `heat pump` vs. `heatpump`. Seems like `heat pump` is the "more correct" spelling.

```suggestion
                        "Connected to Luxtronik heat pump %s:%s", self._host, self._port
```